abci/src/application/kvstore.rs

```diff
@@ -0,0 +1,273 @@
+        ResponseCommit {
+            data: app_hash,
+            retain_height: height - 1,
+        }
+    }
+}
+
+/// Manages key/value store state.
+#[derive(Debug)]
+pub struct KeyValueStoreDriver {
+    store: HashMap<String, String>,
+    height: i64,
+    app_hash: Vec<u8>,
+    cmd_rx: Receiver<Command>,
+}
+
+impl KeyValueStoreDriver {
+    fn new(cmd_rx: Receiver<Command>) -> Self {
+        Self {
+            store: HashMap::new(),
+            height: 0,
+            app_hash: vec![0_u8; MAX_VARINT_LENGTH],
+            cmd_rx,
+        }
+    }
+
+    /// Run the driver in the current thread (blocking).
+    pub fn run(mut self) -> Result<()> {
+        loop {
+            let cmd = self
+                .cmd_rx
+                .recv()
+                .map_err(|e| Error::ChannelRecv(e.to_string()))?;
+            match cmd {
+                Command::GetInfo { result_tx } => {
+                    channel_send(&result_tx, (self.height, self.app_hash.clone()))?
+                }
+                Command::Get { key, result_tx } => {
+                    debug!("Getting value for \"{}\"", key);
+                    channel_send(
+                        &result_tx,
+                        (self.height, self.store.get(&key).map(Clone::clone)),
+                    )?;
+                }
+                Command::Set {
+                    key,
+                    value,
+                    result_tx,
+                } => {
+                    debug!("Setting \"{}\" = \"{}\"", key, value);
+                    channel_send(&result_tx, self.store.insert(key, value))?;
+                }
+                Command::Commit { result_tx } => self.commit(result_tx)?,
+            }
+        }
+    }
+
+    fn commit(&mut self, result_tx: Sender<(i64, Vec<u8>)>) -> Result<()> {
+        // As in the Go-based key/value store, simply encode the number of
+        // items as the "app hash"
+        let mut app_hash = BytesMut::with_capacity(MAX_VARINT_LENGTH);
+        encode_varint(self.store.len() as u64, &mut app_hash);
+        self.app_hash = app_hash.to_vec();
+        self.height += 1;
+        channel_send(&result_tx, (self.height, self.app_hash.clone()))
+    }
+}
+
+#[derive(Debug, Clone)]
+enum Command {
+    /// Get the height of the last commit.
+    GetInfo { result_tx: Sender<(i64, Vec<u8>)> },
+    /// Get the key associated with `key`.
+    Get {
+        key: String,
+        result_tx: Sender<(i64, Option<String>)>,
+    },
+    /// Set the value of `key` to to `value`.
+    Set {
+        key: String,
+        value: String,
+        result_tx: Sender<Option<String>>,
+    },
+    /// Commit the current state of the application, which involves recomputing
+    /// the application's hash.
+    Commit { result_tx: Sender<(i64, Vec<u8>)> },
+}
+
+fn channel_send<T>(tx: &Sender<T>, value: T) -> Result<()> {
+    tx.send(value)
+        .map_err(|e| Error::ChannelSend(e.to_string()))
```

Comment:
Can we hold on to the original error here rather than its string representation?

Reply:
I might be able to do that with `eyre` - I'll try it out :+1: Originally I had trouble with this because [SendError](https://doc.rust-lang.org/std/sync/mpsc/struct.SendError.html) is generic over `T`, which would imply that the error needs to be generic over `T` too. But if the error can be wrapped as a `Box<dyn std::error::Error>` this shouldn't be a problem, I think.

Reply:
Oh I see. Since it's just an example app it does not matter too much, just a nice-to-have (probably not worth it if we have to pull in `eyre`).

Reply:
Here's the conundrum with eyre, as per https://docs.rs/eyre/0.6.5/eyre/struct.Report.html:

> `Report` requires that the error is `Send`, `Sync`, and `'static`.

This would imply that `T` must be `Send`, `Sync` and `'static` too. In this case I'm trying to send a variant of the `Command` enum:

```rust
enum Command {
    /// Get the height of the last commit.
    GetInfo { result_tx: Sender<(i64, Vec<u8>)> },
    /// Get the key associated with `key`.
    Get {
        key: String,
        result_tx: Sender<(i64, Option<String>)>,
    },
    /// Set the value of `key` to to `value`.
    Set {
        key: String,
        value: String,
        result_tx: Sender<Option<String>>,
    },
    /// Commit the current state of the application, which involves recomputing
    /// the application's hash.
    Commit { result_tx: Sender<(i64, Vec<u8>)> },
}
```

`std::sync::mpsc::Sender` is `!Sync`, and I ideally need to be able to send it across the channel in order to get a dedicated response asynchronously.

Perhaps this is a hint that a better overall design is possible? I suppose falling back to `Arc<Mutex<...>>` instead of the handle/driver architecture would solve this. Or simply "flattening" `std::sync::mpsc::SendError<T>`, like I'm doing in the original code here. I can't think why the `SendError` would need to be generic over `T` in the first place?

Reply:
I think the current solution and architecture are fine.

Just wondering, have you tried just casting `SendError` to `Box<dyn std::error::Error>` and wrapping that one in `ChannelSend` (without using `eyre` at all)?

---

Regarding `SendError` being generic over `T`: that's because it's defined as

```rust
pub struct SendError<T>(pub T);
```

which allows you to get back the value that failed to be sent.

Reply:
> Just wondering, have you tried just casting SendError to Box<dyn std::error::Error> and wrapping that one in ChannelSend (without using eyre at all)?

`T` would still need to implement `Sync`, afaik :disappointed: